dex/encrypt/encrypt.go

```diff
@@ -0,0 +1,232 @@
+}
+
+// newArgonPolyCrypter is the constructor for an argonPolyCrypter.
+func newArgonPolyCrypter(pw string) *argonPolyCrypter {
+	pwB := []byte(pw)
+	salt := newSalt()
+	threads := uint8(runtime.NumCPU())
+	keyB := argon2.IDKey(pwB, salt[:], defaultTime, defaultMem, threads, KeySize)
+	var key Key
+	copy(key[:], keyB)
+	return &argonPolyCrypter{
+		key:  key,
+		salt: salt,
+		params: &argonParams{
+			time:    defaultTime,
+			memory:  defaultMem,
+			threads: threads,
+		},
+	}
+}
+
+// Encrypt encrypts the plaintext.
+func (c *argonPolyCrypter) Encrypt(plainText []byte) ([]byte, error) {
+	keyHash := blake256.Sum256(c.key[:])
+	boxer, err := chacha20poly1305.NewX(c.key[:])
+	if err != nil {
+		return nil, fmt.Errorf("aead error: %v", err)
+	}
+	nonceSize := boxer.NonceSize()
+	nonce := make([]byte, nonceSize)
+	_, err = rand.Read(nonce)
+	if err != nil {
+		return nil, fmt.Errorf("nonce generation error: %v", err)
+	}
+	cipherText := boxer.Seal(nil, nonce, plainText, keyHash[:])
+	return encode.BuildyBytes{0}.AddData(nonce).AddData(cipherText), nil
+}
+
+// Decrypt decrypts the ciphertext created by Encrypt.
+func (c *argonPolyCrypter) Decrypt(encrypted []byte) ([]byte, error) {
+	ver, pushes, err := encode.DecodeBlob(encrypted)
+	if err != nil {
+		return nil, fmt.Errorf("DecodeBlob: %v", err)
+	}
+	if ver != 0 {
+		return nil, fmt.Errorf("only version 0 encryptions are known. got version %d", ver)
+	}
+	if len(pushes) != 2 {
+		return nil, fmt.Errorf("expected 2 pushes. got %d", len(pushes))
+	}
+	boxer, err := chacha20poly1305.NewX(c.key[:])
+	if err != nil {
+		return nil, fmt.Errorf("aead error: %v", err)
+	}
+	nonce, cipherText := pushes[0], pushes[1]
+	if len(nonce) != boxer.NonceSize() {
+		return nil, fmt.Errorf("incompatible nonce length. expected %d, got %d", boxer.NonceSize(), len(nonce))
+	}
+	keyHash := blake256.Sum256(c.key[:])
+	plainText, err := boxer.Open(nil, nonce, cipherText, keyHash[:])
+	if err != nil {
+		return nil, fmt.Errorf("aead.Open: %v", err)
+	}
+	return plainText, nil
+}
+
+// Serialize serializes the argonPolyCrypter. It is an error to call Serialize
+// on an un-initialized or a closed argonPolyCrypter.
+func (c *argonPolyCrypter) Serialize() ([]byte, error) {
+	if c.key == zeroKey {
+		return nil, fmt.Errorf("cannot Serialize a closed argonPolyCrypter")
+	}
+	ckSum, err := c.Encrypt([]byte(checkPhrase))
+	if err != nil {
+		return nil, fmt.Errorf("checksum encryption: %v", err)
+	}
+	return c.serializeParams().AddData(ckSum), nil
+}
+
+// Close zeros the key. The argonPolyCrypter is useless after calling close.
+func (c *argonPolyCrypter) Close() {
+	for i := range c.key {
+		c.key[i] = 0
+	}
+}
+
+// serializeParams serializes the KDF parameters and salt.
+func (c *argonPolyCrypter) serializeParams() encode.BuildyBytes {
+	return encode.BuildyBytes{0}.
+		AddData(c.salt[:]).
+		AddData(encode.Uint32Bytes(c.params.time)).
+		AddData(encode.Uint32Bytes(c.params.memory)).
+		AddData([]byte{c.params.threads})
```

Comment:
threads does not need to be recorded. the result of performing argon2.IDKey is the same regardless of how much parallelism you allow it.  callers can determine their own best number based on how many cpus are available.

Reply:
I must be missing something then. This test fails. 

```
func TestCompKeys(t *testing.T) {
  salt := newSalt()
  pwB := []byte("pass")
  id1 := argon2.IDKey(pwB, salt[:], defaultTime, defaultMem, 1, KeySize)
  id2 := argon2.IDKey(pwB, salt[:], defaultTime, defaultMem, 2, KeySize)
  if !bytes.Equal(id1, id2) {
    t.Fatalf("%x != %x", id1, id2)
  }
}
```

The **argon2** docs say 

> IDKey derives a key from the password, salt, and cost parameters

where I'm assuming *cost parameters* are both memory and threads. Are you saying I should be able to change the `threads` parameter without affecting the derived key?

Reply:
Ok, I see my misunderstanding.  I was testing changing GOMAXPROCS, but this doesn't change the result of runtime.NumCPU(), only how many concurrent goroutines the runtime will schedule, and so my keys were always being derived using the same thread counts on the same machine.